internal/persistence/ent/client/preservation_task_test.go

```diff
@@ -154,6 +156,120 @@ func TestCreatePreservationTask(t *testing.T) {
 	}
 }
 
+func TestCreatePreservationTasks(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name    string
+		tasks   func(entc *db.Client) []*datatypes.PreservationTask
+		wantErr string
+	}{
+		{
+			name: "Creates multiple preservation tasks",
+			tasks: func(entc *db.Client) []*datatypes.PreservationTask {
+				pa1, _ := addDBFixtures(t, entc)
+				return []*datatypes.PreservationTask{
+					{
+						TaskID:               "20b8b39c-3642-4138-9dab-b266bfca1e87",
+						Name:                 "Task 1",
+						Status:               enums.PreservationTaskStatusDone,
+						StartedAt:            sql.NullTime{Time: time.Now(), Valid: true},
+						CompletedAt:          sql.NullTime{Time: time.Now().Add(time.Second), Valid: true},
+						Note:                 "Note",
+						PreservationActionID: pa1.ID,
+					},
+					{
+						TaskID:               "9cad7233-58b2-46c4-a2e3-20f892859648",
+						Name:                 "Task 2",
+						Status:               enums.PreservationTaskStatusDone,
+						StartedAt:            sql.NullTime{Time: time.Now(), Valid: true},
+						CompletedAt:          sql.NullTime{Time: time.Now().Add(time.Second), Valid: true},
+						Note:                 "Note",
+						PreservationActionID: pa1.ID,
+					},
+				}
+			},
+		},
+		{
+			name: "Creates multiple preservation tasks exceeding batch size",
+			tasks: func(entc *db.Client) []*datatypes.PreservationTask {
+				pa1, _ := addDBFixtures(t, entc)
+				pts := make([]*datatypes.PreservationTask, 0, 300) // Three batches.
```

Comment:
This seems like a good place to use a generator instead of creating a big slice of PreservationTasks.

Reply:
I need the full slice for assert.DeepEqual - what testing approach did you have in mind with an iterator?

Reply:
Hmm, I didn't realize `CreatePreservationTasks()` returned the full slice of created preservation tasks. I guess my counter question is: what's the advantage of accepting an iterator when you return the full slice of results anyway?  I'm trying to imagine a scenario where you would want to generate 3 tranches of data 100 rows at a time (for example), yield to the results to `CreatePreservationTasks()`, but then get back all three hundred results in one slice. :thinking:

Reply:
Yeah, iterators are quite interesting. Because I'm new to coroutines and generators, I keep referring back to https://research.swtch.com/coro, which I've found really helpful for conceptualizing the mechanics and use-cases of coroutines, but not very confident of my understanding.

To me the batch() function fits well with my concept of a coroutine - the batch function receives data from an iterator until it hits the batch threshold, then yields a tranche of records back to the caller which then does some stuff with that tranch, before resuming the batch coroutine.

I think the use case for `CreatePreservationTasks()` where the iterator parameter is sending batched rows from a MySQL query is an interesting one. I guess in that case the hypothetical `ListPreservationTasks()` would do a SELECT query with a limit on the rows returned, and then would return the results as an iterator, then when it got to the last row in the tranche would return control back to it's caller to SELECT the next tranche of rows?  Meanwhile `CreatePreservationTasks()` would be batching the rows from the iterator and inserting them when it hits the batch limit. I guess my naive expectation at that point is that `CreatePreservationTasks()` would yield back the results from that tranche of inserted rows to it's caller so the caller could do whatever it needs to do with those results (e.g. publish events), then the results could be expired before receiving the next tranche of results (and collected by the garbage collector whenever that ran).  This scenario took a while for me to puzzle through because it's a pretty long chain of coroutines to switch between, with batching on both the SELECT and INSERT operations. :exploding_head:

Reply:
What you're saying reasaones with me as well but I am avoiding to adopt the mental model where we think of iterators as coroutines because that's not how they ended up implementing it - Russ' article explores the broader coroutine space, which differs from go1.23's implementation. Unlike Python (run this [example](https://gist.github.com/sevein/0f26279744f27408711d3143e555be8a) if interested), in Go iterators we have a single stack and there is no pausing/resuming, the compiler is inlining the code (these [examples](https://go.dev/wiki/RangefuncExperiment#can-range-over-a-function-perform-as-well-as-hand-written-loops) are clarifying to me).

> I guess in that case the hypothetical ListPreservationTasks() would do a SELECT query with a limit on the rows returned, and then would return the results as an iterator. When the batching function in ListPreservationTasks() got to the last row in the tranche would return control back to it's caller to SELECT the next tranche of rows. 

Depending on the underlying database driver, I suspect you could leverage server-side cursors so the database can stream the result set in chunks efficiently. See [sqlrange](https://github.com/achille-roussel/sqlrange), the Query function can be a good example. That probably means that you may not need to handle the LIMIT clause manually.

> I guess my naive expectation at that point is that CreatePreservationTasks() would yield back the results from that tranche of inserted rows to it's caller so the caller could do whatever it needs to do with those results (e.g. publish events), then the results could be expired before receiving the next tranche of results (and collected by the garbage collector whenever that ran).

Yeah, that make sense. I chose not to return an iterator, concerned that I wouldn't be able to manage the complexity or that the API would become too cryptic. But it looks like that approach would result in less allocations and less pressure in the garbage collector? But I'll start the simpler version where we just work with slices, then we can explore other options further.

Reply:
> What you're saying reasonates with me as well but I am avoiding to adopt the mental model where we think of iterators as coroutines because that's not how they ended up implementing it - Russ' article explores the broader coroutine space, which differs from go1.23's implementation. Unlike Python (run this [example](https://gist.github.com/sevein/0f26279744f27408711d3143e555be8a) if interested), in Go iterators we have a single stack and there is no pausing/resuming, the compiler is inlining the code (these [examples](https://go.dev/wiki/RangefuncExperiment#can-range-over-a-function-perform-as-well-as-hand-written-loops) are clarifying to me).

Man https://go.dev/wiki/RangefuncExperiment#can-range-over-a-function-perform-as-well-as-hand-written-loops is interesting, but am I the only one that has a really hard time parsing functions used function params or return values? :exploding_head: I had to read those examples at least three times to figure out what was going on. :laughing: 

I'm not really clear on the practical differences between using Go's inlining iterator functions and Python's coroutines, but I did try to reimplement your Python coroutine in Go.  I did get the Go coroutine output to match the Python output, but it was much harder in Go, and I had to use `iter.Pull` to get access to a `next()` function.  The exercise highlighted just how shaky my understanding is on how to write and use iterators in Go. :grimacing: 

> Depending on the underlying database driver, I suspect you could leverage server-side cursors so the database can stream the result set in chunks efficiently. See [sqlrange](https://github.com/achille-roussel/sqlrange), the Query function can be a good example. That probably means that you may not need to handle the LIMIT clause manually. But they're not supported in MySQL as far as I know.

Hmm, server-side cursors sound cool.  We could use whatever method works best for chunking the query results here.

> Yeah, that make sense. I chose not to return an iterator, concerned that I wouldn't be able to manage the complexity or that the API would become too cryptic. But it looks like that approach would result in less allocations and less pressure in the garbage collector? But I'll start the simpler version where we just work with slices, then we can explore other options further.

Doing the simpler version of passing slices sounds good to me. :+1:

Reply:
> I did get the Go coroutine output to match the Python output, but it was much harder in Go, and I had to use iter.Pull to get access to a next() function.

Thanks for doing this, it made me challenge some assumptions I had made. I realized I misunderstood a few things, and this conversation has been helpful in clarifying them. It turns out that Go used [a lightweight/constrained version of coroutines](https://github.com/golang/go/blob/go1.23.3/src/runtime/coro.go) to implement [iter.Pull](https://github.com/golang/go/blob/c390a1c22e8951263e6c01346a4281d604b25062/src/iter/iter.go#L241). This post was very clarifying: https://unskilled.blog/posts/coroutines-in-go/.
I feel like I've finally reached the point where I know I know nothing, so now I can get back to the task I was supposed to be doing in the first place and debug it like a monkey, haha.